In [114]:
import numpy as np
import pandas as pd
import matplotlib as plt
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [115]:
# null hypothesis- one categorical
# from scipy import stats

# F, p = stats.f_oneway(red_wine[red_wine['quality_label'] == 'low']['alcohol'],
# red_wine[red_wine['quality_label'] == 'medium']['alcohol'],
# red_wine[red_wine['quality_label'] == 'high']['alcohol'])
# print('ANOVA test for mean alcohol levels across wine samples with different quality ratings')
# print('F Statistic:', F, '\tp-value:', p)

## Data Preparatrion

In [116]:
white_wine_heatmap = pd.read_csv('winequality-white.csv', sep=';')

In [117]:
white_wine_heatmap

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         8.8        6  
1         9.5        6  
2        10.1        6  
3         9.9        6  
4         9.9        6  
...       ...      ...  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6  

[4898 rows x 12 columns]

In [118]:
white_wine_heatmap.drop_duplicates(inplace=True)
white_wine_heatmap.reset_index(inplace=True)
white_wine_heatmap.drop(columns='index', inplace=True)

In [119]:
white_wine_heatmap.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3961 entries, 0 to 3960
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         3961 non-null   float64
 1   volatile acidity      3961 non-null   float64
 2   citric acid           3961 non-null   float64
 3   residual sugar        3961 non-null   float64
 4   chlorides             3961 non-null   float64
 5   free sulfur dioxide   3961 non-null   float64
 6   total sulfur dioxide  3961 non-null   float64
 7   density               3961 non-null   float64
 8   pH                    3961 non-null   float64
 9   sulphates             3961 non-null   float64
 10  alcohol               3961 non-null   float64
 11  quality               3961 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 371.5 KB


### add quality label

In [120]:
# we are creating a new column called "quality_label", we define a range and associate that range with a label
white_wine_heatmap['quality label'] = white_wine_heatmap['quality'].apply(lambda value: 'low'
if value <= 5 else 'medium'
if value <= 7 else 'high')

# here we are transforming these labels into categrical data type (specific to pandas) instead of simple string
white_wine_heatmap['quality label'] = pd.Categorical(white_wine_heatmap['quality label'],
categories=['low', 'medium', 'high'])

# hypothesis check

In [121]:
from scipy import stats

In [122]:
def anova_tests2(df):  
    
    features = ['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol']
    
    for feature in features:
        F, p = stats.f_oneway(
            df[df['quality label'] == 'low'][feature],
            df[df['quality label'] == 'medium'][feature],
            df[df['quality label'] == 'high'][feature]
        )
        if p > 0.05:
         print(f"ANOVA test for mean {feature} across wine samples with different quality ratings")
         print('F Statistic:', F, '\tp-value:', p)
        else:
         print("there is correlation")

In [123]:
anova_tests2(white_wine_heatmap)


there is correlation
there is correlation
ANOVA test for mean citric acid across wine samples with different quality ratings
F Statistic: 0.24868681586895755 	p-value: 0.7798363477964527
there is correlation
there is correlation
ANOVA test for mean free sulfur dioxide across wine samples with different quality ratings
F Statistic: 0.11569157788750334 	p-value: 0.8907529199895233
there is correlation
there is correlation
there is correlation
there is correlation
there is correlation


In [124]:
# changing wine type to 0 and 1
wine1=white_wine_heatmap.copy()

In [125]:
white_wine_heatmap.shape

(3961, 13)

In [126]:
wine1.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            6.2              0.32         0.16             7.0      0.045   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 30.0                 136.0   0.9949  3.18       0.47   

   alcohol  quality quality label  
0      8.8        6        medium  
1      9.5        6        medium  
2     10.1        6        medium  
3      9.9        6        medium  
4      9.6        6        medium

In [127]:
wine1.shape

(3961, 13)

In [128]:
wine1.drop(columns='citric acid', inplace=True)

In [129]:
wine1.drop(columns='free sulfur dioxide', inplace=True)

## encoding and dropping

In [130]:
from sklearn.preprocessing import OrdinalEncoder
# Create an instance of OrdinalEncoder

In [131]:
ordinal_encoder = OrdinalEncoder(categories=[['low', 'medium', 'high']])
wine1['NumericalLabels'] = ordinal_encoder.fit_transform(wine1[['quality label']])

In [132]:
wine1.drop(columns='quality', inplace=True)

In [133]:
wine1.drop(columns='quality label', inplace=True)

## removing outliers just from alcohol and density

In [134]:
import pandas as pd
import numpy as np

# Assuming df is your pandas DataFrame with features and target
# Define the threshold for outlier detection (e.g., Z-score threshold)
z_threshold = 3

# Select the columns for outlier detection
columns_to_check = ['alcohol', 'density']

# Calculate Z-scores for the selected columns
z_scores = np.abs((wine1[columns_to_check] - wine1[columns_to_check].mean()) / wine1[columns_to_check].std())

# Find outliers based on Z-score threshold
outliers = z_scores > z_threshold

# Remove rows containing outliers
cleaned_df = wine1[~outliers.any(axis=1)]

# Print the shape of the cleaned DataFrame
print("Original DataFrame shape:", wine1.shape)
print("Cleaned DataFrame shape:", cleaned_df.shape)

Original DataFrame shape: (3961, 10)
Cleaned DataFrame shape: (3958, 10)


In [135]:
wine2=cleaned_df.copy()
wine3=cleaned_df.copy()

# Models

divide into train and test

In [136]:
#importm library for dicidint the dataset into two 
from sklearn.model_selection import train_test_split

In [137]:
wine2.head(1)

fixed acidity  volatile acidity  residual sugar  chlorides  \
0            7.0              0.27            20.7      0.045   

   total sulfur dioxide  density   pH  sulphates  alcohol  NumericalLabels  
0                 170.0    1.001  3.0       0.45      8.8              1.0

In [138]:
wine2.drop(columns='NumericalLabels', inplace=True)

In [139]:
X=wine2

In [140]:

y = wine3['NumericalLabels']  # --> what you're trying to predict
  # --> the features we will keep to build our model





In [141]:
X.shape

(3958, 9)

In [142]:
y.shape

(3958,)

In [143]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=100)

In [144]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(3166, 9) (792, 9) (3166,) (792,)


## oversampling

In [145]:
#from imblearn.over_sampling import SMOTE

In [146]:
#from sklearn.datasets import make_classification
#from collections import Counter

In [147]:
# Display class distribution before oversampling
#print("Class Distribution Before SMOTE (Training Set):", Counter(y_train))

In [148]:
# Apply SMOTE for oversampling only on the training set
#smote = SMOTE(sampling_strategy='auto', random_state=42)
#X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

In [149]:
# Display class distribution after oversampling
#print("Class Distribution After SMOTE (Training Set):", Counter(y_train_resampled))

## Random Forest model

## no scaling

In [181]:
from sklearn.datasets import make_classification
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV, StratifiedKFold
from scipy.stats import randint
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report


In [182]:
RF_model= RandomForestClassifier()

In [184]:
# Define the hyperparameter grid for RandomizedSearchCV
random_param_dist = {
    'n_estimators': randint(50, 200),
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': randint(2, 20),
    'min_samples_leaf': randint(1, 10)
}

In [185]:
# Perform RandomizedSearchCV
random_search = RandomizedSearchCV(estimator=RF_model, param_distributions=random_param_dist, 
                                   n_iter=10, scoring='accuracy', cv=5, random_state=42)
random_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(),
                   param_distributions={'max_depth': [None, 10, 20, 30],
                                        'min_samples_leaf': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x00000248E58E1690>,
                                        'min_samples_split': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x00000248E5958B10>,
                                        'n_estimators': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x00000248E5958E10>},
                   random_state=42, scoring='accuracy')

In [191]:
# Get the best hyperparameters from RandomizedSearchCV
best_random_params = random_search.best_params_
print("Best Hyperparameters from RandomizedSearchCV:", best_random_params)

Best Hyperparameters from RandomizedSearchCV: {'max_depth': None, 'min_samples_leaf': 3, 'min_samples_split': 13, 'n_estimators': 104}


In [190]:
# Define the hyperparameter grid for GridSearchCV around the best parameters found by RandomizedSearchCV
param_grid = {
    'n_estimators': [best_random_params['n_estimators'] - 20, best_random_params['n_estimators'], 
                     best_random_params['n_estimators'] + 20],
    'max_depth': [None] + [best_random_params['max_depth']],
    'min_samples_split': [best_random_params['min_samples_split'] - 2, 
                          best_random_params['min_samples_split'], 
                          best_random_params['min_samples_split'] + 2],
    'min_samples_leaf': [best_random_params['min_samples_leaf'] - 1, 
                         best_random_params['min_samples_leaf'], 
                         best_random_params['min_samples_leaf'] + 1]
}

In [189]:
# Define stratified k-fold cross-validation
stratified_kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [192]:
# Perform GridSearchCV
grid_search = GridSearchCV(estimator=RF_model, param_grid=param_grid, cv=stratified_kfold)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=42, shuffle=True),
             estimator=RandomForestClassifier(),
             param_grid={'max_depth': [None, None],
                         'min_samples_leaf': [2, 3, 4],
                         'min_samples_split': [11, 13, 15],
                         'n_estimators': [84, 104, 124]})

In [174]:


# Get the best hyperparameters from GridSearchCV
best_params = grid_search.best_params_
print("Best Hyperparameters from GridSearchCV:", best_params)

# Train the final model with the best hyperparameters
best_rf_classifier = RandomForestClassifier(**best_params)
best_rf_classifier.fit(X_train, y_train)

# Evaluate the final model on the testing set
test_accuracy = best_rf_classifier.score(X_test, y_test)
print("Test Accuracy:", test_accuracy)

In [175]:
stratified_kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [176]:
# Perform grid search with cross-validation
grid_search = GridSearchCV(estimator=RF_model, param_grid=param_grid, cv=stratified_kfold)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=42, shuffle=True),
             estimator=RandomForestClassifier(),
             param_grid={'max_depth': [None, 10, 20],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [50, 100, 200]})

In [193]:
# Get the best hyperparameters
best_params = grid_search.best_params_
print("Best Hyperparameters:", best_params)

Best Hyperparameters: {'max_depth': None, 'min_samples_leaf': 3, 'min_samples_split': 15, 'n_estimators': 124}


In [195]:
# Evaluate the model with best hyperparameters on the test set
best_rf_classifier = RandomForestClassifier(**best_params)
best_rf_classifier.fit(X_train, y_train)
y_pred = best_rf_classifier.predict(X_test)

In [196]:
# Print classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

         0.0       0.70      0.61      0.65       282
         1.0       0.76      0.85      0.80       487
         2.0       0.00      0.00      0.00        23

    accuracy                           0.74       792
   macro avg       0.49      0.49      0.49       792
weighted avg       0.72      0.74      0.73       792



c:\Users\gabb_\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\gabb_\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\gabb_\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### stratified fold

from sklearn.model_selection import StratifiedKFold, cross_val_score

# Create a StratifiedKFold object with k=5 (you can adjust the number of folds)
stratified_kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
cross_val_scores = cross_val_score(RF_model, X, y, cv=stratified_kfold, scoring='accuracy')

# Print the cross-validation scores
print("Cross-Validation Scores:")
print(cross_val_scores)

# Print the mean and standard deviation of the cross-validation scores
print(f"\nMean Accuracy: {cross_val_scores.mean():.2f}")
print(f"Standard Deviation: {cross_val_scores.std():.2f}")

RF_model.fit(X_train, y_train)

In [154]:
preds2= RF_model.predict(X_test)

In [155]:
preds2

array([1., 0., 0., 0., 0., 0., 1., 1., 0., 1., 1., 0., 0., 1., 0., 1., 1.,
       1., 1., 0., 1., 1., 0., 0., 1., 0., 1., 0., 1., 0., 0., 1., 1., 1.,
       1., 1., 0., 0., 1., 0., 1., 1., 1., 1., 1., 1., 0., 0., 1., 1., 1.,
       1., 0., 1., 1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 0., 1.,
       0., 1., 1., 1., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1., 0., 0.,
       0., 1., 0., 1., 1., 1., 1., 1., 1., 0., 0., 1., 1., 0., 1., 1., 1.,
       1., 0., 1., 1., 1., 0., 1., 1., 1., 1., 0., 1., 0., 0., 1., 0., 0.,
       0., 1., 1., 1., 1., 1., 0., 1., 1., 1., 0., 1., 1., 0., 1., 1., 0.,
       1., 1., 1., 1., 1., 0., 0., 1., 1., 1., 1., 1., 0., 1., 1., 1., 0.,
       1., 1., 1., 1., 1., 0., 1., 1., 0., 1., 0., 1., 0., 1., 1., 1., 1.,
       1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0., 1., 0.,
       0., 1., 1., 1., 0., 1., 1., 1., 0., 1., 1., 1., 1., 0., 0., 1., 1.,
       0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 1.,
       1., 1., 0., 0., 0.

In [197]:
#acc=accuracy_score(y_test, preds2)
#acc

## results

In [157]:
# from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [198]:
# accuracy = accuracy_score(y_test, preds2)
# print(f"Accuracy: {accuracy}")

# print("\nClassification Report:")
# print(classification_report(y_test, preds2))

# print("\nConfusion Matrix:")
# print(confusion_matrix(y_test, preds2))

1st try no scaling or oversampling
Accuracy: 0.7301387137452712

Classification Report:
              precision    recall  f1-score   support

         0.0       0.67      0.63      0.65       274
         1.0       0.76      0.83      0.79       490
         2.0       0.50      0.03      0.06        29

    accuracy                           0.73       793
   macro avg       0.64      0.50      0.50       793
weighted avg       0.72      0.73      0.72       793


Confusion Matrix:
[[172 102   0]
 [ 83 406   1]
 [  0  28   1]]

2nd try oversampling:
Accuracy: 0.669609079445145

Classification Report:
              precision    recall  f1-score   support

         0.0       0.61      0.70      0.65       274
         1.0       0.77      0.68      0.72       490
         2.0       0.12      0.17      0.14        29

    accuracy                           0.67       793
   macro avg       0.50      0.52      0.50       793
weighted avg       0.69      0.67      0.68       793


Confusion Matrix:
[[193  79   2]
 [121 333  36]
 [  2  22   5]]

3rd try removed irrelevant columns:
Accuracy: 0.6620428751576293

Classification Report:
              precision    recall  f1-score   support

         0.0       0.60      0.70      0.65       274
         1.0       0.77      0.66      0.71       490
         2.0       0.17      0.31      0.22        29

    accuracy                           0.66       793
   macro avg       0.51      0.56      0.53       793
weighted avg       0.69      0.66      0.67       793


Confusion Matrix:
[[191  79   4]
 [124 325  41]
 [  1  19   9]]


4th try: removed the oversampling
Accuracy: 0.7112232030264817

Classification Report:
              precision    recall  f1-score   support

         0.0       0.65      0.59      0.62       274
         1.0       0.74      0.82      0.78       490
         2.0       0.50      0.03      0.06        29

    accuracy                           0.71       793
   macro avg       0.63      0.48      0.49       793
weighted avg       0.70      0.71      0.70       793


Confusion Matrix:
[[162 112   0]
 [ 88 401   1]
 [  1  27   1]]

5th try: removing outliers from two columns:
Accuracy: 0.7361111111111112

Classification Report:
              precision    recall  f1-score   support

         0.0       0.70      0.59      0.64       282
         1.0       0.75      0.85      0.80       487
         2.0       1.00      0.04      0.08        23

    accuracy                           0.74       792
   macro avg       0.82      0.50      0.51       792
weighted avg       0.74      0.74      0.72       792


Confusion Matrix:
[[167 115   0]
 [ 72 415   0]
 [  1  21   1]]


6th try: stratified fold:
Accuracy: 0.7361111111111112

Classification Report:
              precision    recall  f1-score   support

         0.0       0.70      0.59      0.64       282
         1.0       0.75      0.85      0.80       487
         2.0       1.00      0.04      0.08        23

    accuracy                           0.74       792
   macro avg       0.82      0.50      0.51       792
weighted avg       0.74      0.74      0.72       792


Confusion Matrix:
[[167 115   0]
 [ 72 415   0]
 [  1  21   1]]

## KNN

In [159]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [160]:
scaler = StandardScaler()
X_train1 = scaler.fit_transform(X_train)
X_test1 = scaler.transform(X_test)

In [161]:
k = 3  # You can adjust this value based on your problem
knn_model = KNeighborsClassifier(n_neighbors=k)


In [162]:
from sklearn.model_selection import StratifiedKFold, cross_val_score

# Create a StratifiedKFold object with k=5 (you can adjust the number of folds)
stratified_kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
cross_val_scores = cross_val_score(knn_model, X, y, cv=stratified_kfold, scoring='accuracy')

# Print the cross-validation scores
print("Cross-Validation Scores:")
print(cross_val_scores)

# Print the mean and standard deviation of the cross-validation scores
print(f"\nMean Accuracy: {cross_val_scores.mean():.2f}")
print(f"Standard Deviation: {cross_val_scores.std():.2f}")

Cross-Validation Scores:
[0.60227273 0.64141414 0.63131313 0.62326169 0.62958281]

Mean Accuracy: 0.63
Standard Deviation: 0.01


In [163]:
knn_model.fit(X_train1, y_train)

KNeighborsClassifier(n_neighbors=3)

## results


In [164]:
y_pred = knn_model.predict(X_test1)

In [165]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.6919191919191919

Classification Report:
              precision    recall  f1-score   support

         0.0       0.62      0.59      0.60       282
         1.0       0.74      0.78      0.76       487
         2.0       0.14      0.04      0.07        23

    accuracy                           0.69       792
   macro avg       0.50      0.47      0.48       792
weighted avg       0.68      0.69      0.68       792


Confusion Matrix:
[[166 116   0]
 [100 381   6]
 [  3  19   1]]


1st try
Accuracy: 0.6973518284993695

Classification Report:
              precision    recall  f1-score   support

         0.0       0.62      0.62      0.62       274
         1.0       0.75      0.77      0.76       490
         2.0       0.25      0.14      0.18        29

    accuracy                           0.70       793
   macro avg       0.54      0.51      0.52       793
weighted avg       0.69      0.70      0.69       793


Confusion Matrix:
[[171 101   2]
 [102 378  10]
 [  2  23   4]]

2nd try:
Accuracy: 0.6015132408575031

Classification Report:
              precision    recall  f1-score   support

         0.0       0.59      0.74      0.66       274
         1.0       0.78      0.54      0.64       490
         2.0       0.07      0.24      0.10        29

    accuracy                           0.60       793
   macro avg       0.48      0.51      0.47       793
weighted avg       0.69      0.60      0.63       793


Confusion Matrix:
[[204  59  11]
 [136 266  88]
 [  5  17   7]]

3rd try:

Accuracy: 0.562421185372005

Classification Report:
              precision    recall  f1-score   support

         0.0       0.54      0.66      0.60       274
         1.0       0.74      0.52      0.61       490
         2.0       0.09      0.38      0.15        29

    accuracy                           0.56       793
   macro avg       0.46      0.52      0.45       793
weighted avg       0.65      0.56      0.59       793


Confusion Matrix:
[[180  76  18]
 [147 255  88]
 [  4  14  11]]

4th try:
Accuracy: 0.6733921815889029

Classification Report:
              precision    recall  f1-score   support

         0.0       0.58      0.59      0.59       274
         1.0       0.73      0.76      0.74       490
         2.0       0.14      0.03      0.06        29

    accuracy                           0.67       793
   macro avg       0.49      0.46      0.46       793
weighted avg       0.66      0.67      0.66       793


Confusion Matrix:
[[161 113   0]
 [112 372   6]
 [  3  25   1]]

6th try: stratified kfold:
Accuracy: 0.6919191919191919

Classification Report:
              precision    recall  f1-score   support

         0.0       0.62      0.59      0.60       282
         1.0       0.74      0.78      0.76       487
         2.0       0.14      0.04      0.07        23

    accuracy                           0.69       792
   macro avg       0.50      0.47      0.48       792
weighted avg       0.68      0.69      0.68       792


Confusion Matrix:
[[166 116   0]
 [100 381   6]
 [  3  19   1]]

import numpy as np
import matplotlib.pyplot as plt
from sklearn.datasets import make_classification
from sklearn.neighbors import KNeighborsClassifier

# Generate synthetic data with three classes
X, y = make_classification(n_samples=1000, n_features=2, n_classes=3, 
                           n_clusters_per_class=1, n_redundant=0, 
                           n_informative=2, class_sep=1.0, 
                           random_state=42)

# Train KNN classifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X, y)

# Plot decision boundary
x_min, x_max = X[:, 0].min() - 1, X[:, 0].max() + 1
y_min, y_max = X[:, 1].min() - 1, X[:, 1].max() + 1
xx, yy = np.meshgrid(np.arange(x_min, x_max, 0.02),
                     np.arange(y_min, y_max, 0.02))

Z = knn.predict(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)

plt.figure(figsize=(8, 6))
plt.contourf(xx, yy, Z, alpha=0.8)
plt.scatter(X[:, 0], X[:, 1], c=y, marker='o', edgecolors='k')
plt.xlabel('Feature 1')
plt.ylabel('Feature 2')
plt.title('Decision Boundary Plot for KNN (Three Classes)')
plt.colorbar()
plt.show()

## Support Vector Machines (SVM)

In [166]:
# Import necessary libraries
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report


In [167]:

# Standardize the features (important for SVM)
scaler = StandardScaler()
X4_train = scaler.fit_transform(X_train)
X4_test = scaler.transform(X_test)

In [168]:
# Initialize the SVM classifier
svm_classifier = SVC(kernel='linear', C=1.0, random_state=42)

In [169]:


# Train the SVM classifier
svm_classifier.fit(X4_train, y_train)

# Make predictions on the test set
prediction4 = svm_classifier.predict(X4_test)

## RESULT

In [170]:
# Evaluate the performance of the classifier
accuracy4 = accuracy_score(y_test, prediction4)
report4 = classification_report(y_test, prediction4)

# Print the results
print(f"Accuracy: {accuracy4:.2f}")
print("\nClassification Report:\n", report4)
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, prediction4))

Accuracy: 0.70

Classification Report:
               precision    recall  f1-score   support

         0.0       0.65      0.51      0.57       282
         1.0       0.72      0.84      0.77       487
         2.0       0.00      0.00      0.00        23

    accuracy                           0.70       792
   macro avg       0.46      0.45      0.45       792
weighted avg       0.67      0.70      0.68       792


Confusion Matrix:
[[145 137   0]
 [ 79 408   0]
 [  0  23   0]]


c:\Users\gabb_\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\gabb_\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\gabb_\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


1st try:
Accuracy: 0.70

Classification Report:
               precision    recall  f1-score   support

         0.0       0.64      0.53      0.58       274
         1.0       0.72      0.83      0.77       490
         2.0       0.00      0.00      0.00        29

    accuracy                           0.70       793
   macro avg       0.45      0.45      0.45       793
weighted avg       0.67      0.70      0.68       793


Confusion Matrix:
[[145 129   0]
 [ 82 408   0]
 [  0  29   0]]

2nd try:
Accuracy: 0.52

Classification Report:
               precision    recall  f1-score   support

         0.0       0.56      0.73      0.64       274
         1.0       0.76      0.38      0.51       490
         2.0       0.11      0.72      0.19        29

    accuracy                           0.52       793
   macro avg       0.48      0.61      0.45       793
weighted avg       0.67      0.52      0.54       793


Confusion Matrix:
[[201  52  21]
 [156 188 146]
 [  0   8  21]]

3rd try:
Accuracy: 0.51

Classification Report:
               precision    recall  f1-score   support

         0.0       0.56      0.74      0.64       274
         1.0       0.73      0.37      0.49       490
         2.0       0.11      0.69      0.18        29

    accuracy                           0.51       793
   macro avg       0.47      0.60      0.44       793
weighted avg       0.65      0.51      0.53       793


Confusion Matrix:
[[202  56  16]
 [158 179 153]
 [  0   9  20]]

4th try:
Accuracy: 0.69

Classification Report:
               precision    recall  f1-score   support

         0.0       0.62      0.51      0.56       274
         1.0       0.71      0.82      0.77       490
         2.0       0.00      0.00      0.00        29

    accuracy                           0.69       793
   macro avg       0.44      0.45      0.44       793
weighted avg       0.66      0.69      0.67       793


Confusion Matrix:
[[141 133   0]
 [ 86 404   0]
 [  0  29   0]]